File: misli-core/misli/basic_classes/point2d.py
import math


class Point2D:
    def __init__(self, x: float, y: float):
        self._x = x
        self._y = y

    @classmethod
    def from_coords(cls, coords: list):
        if len(coords) != 2:
            raise ValueError

        x = coords[0]
        y = coords[1]

        return cls(x, y)

    def __repr__(self):
        return '<Point x=%s y=%s>' % (self.x(), self.y())

    def __add__(self, other: 'Point2D'):
        return Point2D(self.x() + other.x(), self.y() + other.y())

    def __sub__(self, other: 'Point2D'):
        return Point2D(self.x() - other.x(), self.y() - other.y())

    def __truediv__(self, k):
        return Point2D(self.x() / k, self.y() / k)

    def x(self):
        return self._x

    def y(self):
        return self._y

    def to_list(self):
        return [self._x, self._y]

    def distance_to(self, point: 'Point2D'):
        distance = math.sqrt(
            (self.x() - point.x())**2 + (self.y() - point.y())**2)
        return distance

File: web-app/src/util/Point2D.ts
// import { Point2D } from '../types/util';
export type PointData = [number, number];

export class Vector2D {
  constructor(public x: number = 0, public y: number = 0) { }

  equals(other: Vector2D): boolean {
    return this.x === other.x && this.y === other.y;
  }

  add(other: Vector2D): this {
    return new (this.constructor as any)(this.x + other.x, this.y + other.y);
  }

  subtract(other: Vector2D): this {
    return new (this.constructor as any)(this.x - other.x, this.y - other.y);
  }

  divide(k: number): this {
    return new (this.constructor as any)(this.x / k, this.y / k);
  }

  round(): this {
    return new (this.constructor as any)(Math.round(this.x), Math.round(this.y));
  }

  multiply(k: number): this {
    return new (this.constructor as any)(this.x * k, this.y * k);
  }
}

export class Point2D  extends Vector2D {

  static fromData(obj: PointData): Point2D {
    return new Point2D(obj[0], obj[1]);
  }

  data(): PointData {
    return [this.x, this.y];
  }

  copy(): Point2D {
    return new Point2D(this.x, this.y);
  }

  toString(): string {
    return `<Point x=${this.x} y=${this.y}>`;
  }

  distanceTo(point: Point2D): number {
    const distance = Math.sqrt(
      Math.pow(this.x - point.x, 2) + Math.pow(this.y - point.y, 2)
    );
    return distance;
  }

  rotated(radians: number, origin: Point2D): Point2D {
    const adjustedX = this.x - origin.x;
    const adjustedY = this.y - origin.y;
    const cosRad = Math.cos(radians);
    const sinRad = Math.sin(radians);
    const qx = origin.x + cosRad * adjustedX + sinRad * adjustedY;
    const qy = origin.y - sinRad * adjustedX + cosRad * adjustedY;
    return new Point2D(qx, qy);
  }
}
